### ISSUE TITLE
Per-release sourceUrl support for npm datasource

### ISSUE THREAD

**Issue by KnisterPeter** [renovatebot/renovate#13943]
### What would you like Renovate to be able to do?

I would like renovate select the sourceUrl per release.
This is helpful if a package moves the source repository (e.g. from github to gitlab).

### If you have any ideas on how this should be implemented, please tell us here.

Currently renovate selects the global repository of a package over the release specific repository:

https://github.com/renovatebot/renovate/blob/790ff0196f75900da4b80161e0990b32d3406a39/lib/datasource/npm/get.ts#L101

Not sure if it is enough to switch order here and prioritize the pre-release repository over the package repository.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by KnisterPeter** [comment]
This issue was created of a discussion over here: https://github.com/renovatebot/renovate/discussions/13940

**Comment by rarkins** [comment]
> Currently renovate selects the global repository of a package over the release specific repository:
> 
> https://github.com/renovatebot/renovate/blob/790ff0196f75900da4b80161e0990b32d3406a39/lib/datasource/npm/get.ts#L101
> 
> Not sure if it is enough to switch order here and prioritize the pre-release repository over the package repository.

No, that would be undesirable. The package.sourceUrl value should remain as-is. What you need is for release.sourceUrl to be added if it differs from package.sourceUrl.

**Comment by rarkins** [comment]
Need to find a public npm package to test this against, which has different repositories in certain releases

**Comment by ZauberNerd** [comment]
@rarkins I did a quick test with npmjs, check out version 8 and 9 of: https://www.npmjs.com/package/zaubernerd-test-package
```shell
curl https://registry.npmjs.org/zaubernerd-test-package | jq
```
As we can see here, npm does respect the updated repository url, while other registries (such as Nexus) might not.

**Comment by KnisterPeter** [comment]
If we can group PRs by `release.sourceUrl` then it would be fine.

**Comment by rarkins** [comment]
@ZauberNerd thanks. This feature primarily needed as a workaround for incorrect Nexus behavior, but it's still a good thing to have. e.g. let's say someone was updating from v7 to v8 of your package, then we should show the "old" sourceUrl then and not the new one.
